src/file-events/cpp/linux_fsnotifier.cpp

```diff
@@ -218,28 +251,44 @@ void Server::handleEvent(JNIEnv* env, const inotify_event* event) {
     reportChange(env, type, path);
 }
 
+static int addInotifyWatch(const u16string& path, shared_ptr<Inotify> inotify) {
+    string pathNarrow = utf16ToUtf8String(path);
+    int fdWatch = inotify_add_watch(inotify->fd, pathNarrow.c_str(), EVENT_MASK);
+    if (fdWatch == -1) {
+        throw FileWatcherException("Couldn't add watch", path, errno);
+    }
+    return fdWatch;
+}
+
 void Server::registerPath(const u16string& path) {
-    if (watchPoints.find(path) != watchPoints.end()) {
+    auto it = watchPoints.find(path);
+    if (it != watchPoints.end()) {
+        auto& watchPoint = it->second;
+        if (watchPoint.status != FINISHED) {
```

Comment:
We probably should removed the `FINISHED` nodes at some point from the `watchPoints`, so we can free up the memory.

This is currently not a problem since we stop the watchers every build.